Rubberduck.VBEEditor/ComManagement/ProjectsRepository.cs

```diff
@@ -0,0 +1,267 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using Rubberduck.VBEditor.SafeComWrappers.Abstract;
+using Rubberduck.VBEditor.Extensions;
+using Rubberduck.VBEditor.SafeComWrappers;
+
+namespace Rubberduck.VBEditor.ComManagement
+{
+    public class ProjectsRepository : IProjectsRepository
+    {
+        private IVBProjects _projectsCollection;
+        private readonly IDictionary<string, IVBProject> _projects = new Dictionary<string, IVBProject>();
+        private readonly IDictionary<string, IVBComponents> _componentsCollections = new Dictionary<string, IVBComponents>();
+        private readonly IDictionary<QualifiedModuleName, IVBComponent> _components = new Dictionary<QualifiedModuleName, IVBComponent>();
+
+        private readonly ReaderWriterLockSlim _refreshProtectionLock = new ReaderWriterLockSlim(LockRecursionPolicy.NoRecursion);
+
+        public ProjectsRepository(IVBE vbe)
+        {
+            _projectsCollection = vbe.VBProjects;
+        }
+
+        private void LoadCollections()
+        {
+            LoadProjects();
+            LoadComponentsCollections();
+            LoadComponents();
+        }
+
+        private void LoadProjects()
+        {
+            foreach (var project in _projectsCollection)
+            {
+                if (project.Protection == ProjectProtection.Locked)
+                {
+                    project.Dispose();
+                    continue;
+                }
+
+                EnsureValidProjectId(project);
+                _projects.Add(project.ProjectId, project);
+            }
+        }
+
+        private void EnsureValidProjectId(IVBProject project)
+        {
+            if (string.IsNullOrEmpty(project.ProjectId) || _projects.Keys.Contains(project.ProjectId))
+            {
+                project.AssignProjectId();
+            }
+        }
+
+        private void LoadComponentsCollections()
+        {
+            foreach (var (projectId, project) in _projects)
+            {
+                _componentsCollections.Add(projectId, project.VBComponents);
+            }
+        }
+
+        private void LoadComponents()
+        {
+            foreach (var components in _componentsCollections.Values)
+            {
+                LoadComponents(components);
+            }
+        }
+
+        private void LoadComponents(IVBComponents componentsCollection)
+        {
+            foreach (var component in componentsCollection)
+            {
+                var qmn = component.QualifiedModuleName;
+                _components.Add(qmn, component);
+            }
+        }
+
+        public void Refresh()
+        {
+            ExecuteWithinWriteLock(() => RefreshCollections());
+        }
+
+        private void ExecuteWithinWriteLock(Action action)
+        {
+            if (_disposed)
+            {
+                return; //The lock has already been diposed.
```

Comment:
Typo in the comment: should be `disposed`

Reply:
thanks, will fix